src/Datasource/EntityTrait.php

```diff
@@ -1036,12 +1036,58 @@ protected function _readError($object, $path = null)
     }
 
     /**
+     * Get a list of invalid fields and their data for errors upon validation/patching
+     *
+     * @return array
+     */
+    public function getInvalid()
+    {
+        return $this->_invalid;
+    }
+
+    /**
+     * Sets a field as invalid and not patchable into the entity.
+     *
+     * This is useful for batch operations when one needs to get the original value for an error message after patching.
+     * This value could not be patched into the entity and is simply copied into the _invalid property for debugging purposes
+     * or to be able to log it away.
+     *
+     * @param string|array $field The value to set.
+     * @param mixed|null $value The invalid value to be set for $field.
+     * @param bool $overwrite Whether or not to overwrite pre-existing values for $field.
+     * @return $this|mixed
+     */
+    public function setInvalid($field, $value = null, $overwrite = false)
```

Comment:
@cleptric changing the method behavior based on argument type can be confusing. Also this would make a better transition to PHP7:
```php
public function setInvalid(array $fields): self;
public function setInvalidField(string $field, string $value): self
```